### Day 08: Boot Code

In [1]:

# read & clean input data 
with open('day08_test_data.txt') as fh:
    lines = [line.strip() for line in fh.readlines()] 
    
# split into appropriate values
clean_commands = [(x.split(" ")[0], int(x.split(" ")[1])) for x in lines]

In [2]:
accumulator = 0
step = 0
track_steps = [0]
keep = True
while keep:
    command = clean_commands[step][0]
    if command == "acc":
        accumulator += clean_commands[step][1]
        step += 1
    elif command == "jmp":
        step += clean_commands[step][1]
    elif command == "nop":
        step += 1
    
    # track next step & determine if visited already
    if step in track_steps:
        print(accumulator)
        break
        # otherwise move on 
    track_steps.append(step)
    print(track_steps)
        
    

### Real Data

In [3]:
# read & clean input data 
with open('day08_data.txt') as fh:
    lines = [line.strip() for line in fh.readlines()] 
    
# split into appropriate values
clean_commands = [(x.split(" ")[0], int(x.split(" ")[1])) for x in lines]
len(clean_commands)

660

In [4]:
accumulator = 0
step = 0
track_steps = [0]
keep = True
while keep:
    command = clean_commands[step][0]
    if command == "acc":
        accumulator += clean_commands[step][1]
        step += 1
    elif command == "jmp":
        step += clean_commands[step][1]
    elif command == "nop":
        step += 1
    
    # track next step & determine if visited already
    if step in track_steps:
        print(accumulator)
        break
        # otherwise move on 
    track_steps.append(step)

1915


### Part 2:

- Handling a bug -> how to do this in the non-brute force way given we have 660 input lines....

In [5]:
# read & clean input data 
with open('day08_test_data.txt') as fh:
    lines = [line.strip() for line in fh.readlines()] 
    
# split into appropriate values
commands = [x.split(" ")[0] for x in lines]
deltas = [int(x.split(" ")[1]) for x in lines]

In [6]:
# build out possible changes -> only can impact nop or jmp
change_index = [i for i in range(len(commands)) if commands[i] in ["jmp", "nop"]]
print(change_index)

[0, 2, 4, 7]


In [7]:
for idx in change_index:
    accumulator = 0
    step = 0
    track_steps = [0]
    keep = True
    fail_flag = False
    
    while keep and step < len(commands):
        command = commands[step]
        if step == idx:
            if command == "jmp":
                command = "nop"
            else:
                command = "jmp"
        
        # proceed
        if command == "acc":
            accumulator += deltas[step]
            step += 1
        elif command == "jmp":
            step += deltas[step]
        elif command == "nop":
            step += 1
        
        # still track next step & determine if visited already
        if step in track_steps:
            fail_flag = True
            break
            # otherwise move on 
        track_steps.append(step)
    if not fail_flag:
        print(accumulator)
        break

8


### Trying on Real Data

- By using break on repeated steps and adding a flag I can reduce general runtime

In [8]:
# read & clean input data 
with open('day08_data.txt') as fh:
    lines = [line.strip() for line in fh.readlines()] 
    
# split into appropriate values
commands = [x.split(" ")[0] for x in lines]
deltas = [int(x.split(" ")[1]) for x in lines]

# build out possible changes -> only can impact nop or jmp
change_index = [i for i in range(len(commands)) if commands[i] in ["jmp", "nop"]]
print(len(change_index)) # not too many iterations...will take some time 

311


In [9]:
for idx in change_index:
    accumulator = 0
    step = 0
    track_steps = [0]
    keep = True
    fail_flag = False
    
    while keep and step < len(commands):
        command = commands[step]
        if step == idx:
            if command == "jmp":
                command = "nop"
            else:
                command = "jmp"
        
        # proceed
        if command == "acc":
            accumulator += deltas[step]
            step += 1
        elif command == "jmp":
            step += deltas[step]
        elif command == "nop":
            step += 1
        
        # still track next step & determine if visited already
        if step in track_steps:
            fail_flag = True
            break
            # otherwise move on 
        track_steps.append(step)
    if not fail_flag:
        print(accumulator)
        break

944
